File: 2022/aoc_python/day_3.py
from lib import io

print("Advent of Code 2022 Day 3")
print("-------------------------")
rucksacks = io.file_to_list("input/day-3-input.txt")
# print(f"rucksacks: {rucksacks}")

priorities = {chr(k): v for (k, v) in zip(range(97, 123), range(1, 27))}        # Lowercase letters mapped to priority values from 1 to 26.
priorities.update({chr(k): v for (k, v) in zip(range(65, 91), range(27, 53))})  # Uppercase letters mapped to priority values from 27 to 52.

total_priorities = 0
for rucksack in rucksacks:
    # Get all items in common between both compartments.
    compartment_1 = rucksack[:len(rucksack) // 2]
    compartment_2 = rucksack[len(rucksack) // 2:]
    # print(f"compartment_1: {compartment_1}")
    # print(f"compartment_2: {compartment_2}")
    common_items = set(item for item in compartment_1 if item in compartment_2)
    # print(f"common_items: {common_items}")

    # Calculate total priority of common items.
    total_priorities += sum(priorities[item] for item in common_items)

print("PART 1")
print("======")
print(f"Total priorities: {total_priorities}")
print("======")
# Attempt 1: 10716 (too high)
# Attempt 2: 7917 (correct)

# rucksacks = io.file_to_list("input/day-3-test-data.txt")
total_priorities = 0
num_iterations = 0
# i = 3
i = 0
# while i <= len(rucksacks):
while i <= len(rucksacks) - 3:
    # print(f"i: {i}")
    # Get all items in common between the rucksacks, grouped in threes.
    # common_items = set(rucksacks[i - 3]) & set(rucksacks[i - 2]) & set(rucksacks[i - 1])
    common_items = set(rucksacks[i]) & set(rucksacks[i + 1]) & set(rucksacks[i + 2])
    # print(f"common_items: {common_items}")

    # Calculate total priority of common items.
    total_priorities += sum(priorities[item] for item in common_items)

    i += 3
    num_iterations += 1

# print(f"len(rucksacks): {len(rucksacks)}")
# print(f"num_iterations: {num_iterations}")
# print(f"i: {i}")

print("PART 2")
print("======")
print(f"Total priorities: {total_priorities}")
print("======")
# Attempt 1: 2566 (too low)
# Attempt 2: 2585 (correct)

File: 2022/aoc_perl/day_3.pl
#!/usr/bin/perl
use strict;
use warnings;

use FindBin;                                # Useful for finding directory where script is being invoked.
use lib "$FindBin::Bin/../aoc_perl/lib";    # Identifies library folder.

use set qw(make_set union intersection);
use util qw(file_to_array println print_hash);

sub make_priority_hash {
    my %args = (
        STARTING_ASCII_VALUE => 0,
        PRIORITY_START => 0,
        PRIORITY_END => 0,
        @_,
    );
    my $char_value = $args{STARTING_ASCII_VALUE};
    my $priority_start = $args{PRIORITY_START};
    my $priority_end = $args{PRIORITY_END};

    my %priorities = ();
    my $priority_value;
    for ($priority_value = $priority_start; $priority_value <= $priority_end; $priority_value++, $char_value++) {
        $priorities{chr($char_value)} = $priority_value;
    }

    return %priorities;
}

sub make_rucksack_set {
    my %args = (
        RUCKSACK_STR => "",
        @_
    );
    my $rucksack_str = $args{RUCKSACK_STR};

    my $elements = [];
    my $i;
    for ($i = 0; $i < length($rucksack_str); $i++) {
        push(@{$elements}, substr($rucksack_str, $i, 1));
    }
    return make_set(ELEMENTS => $elements);
}

println("Advent of Code 2022 Day 3");
println("-------------------------");

my @rucksacks = file_to_array(
    PATH_TO_FILE => "$FindBin::Bin/../input",
    FILE_NAME => "day-3-input.txt"
);

my %priorities = ();    # For easy access to priority values for characters.

# Lowercase letters mapped to priority values from 1 to 26.
%priorities = make_priority_hash(
    STARTING_ASCII_VALUE => 97, # ASCII/Unicode value for 'a'.
    PRIORITY_START => 1,
    PRIORITY_END => 26
);

# Uppercase letters mapped to priority values from 27 to 52.
%priorities = (
    %priorities,
    make_priority_hash(
        STARTING_ASCII_VALUE => 65, # ASCII/Unicode value for 'A'.
        PRIORITY_START => 27,
        PRIORITY_END => 52
    )
);

my $total_priorities = 0;
foreach my $rucksack (@rucksacks) {
    use integer;
    my $compartment1 = substr($rucksack, 0, length($rucksack) / 2);
    my $compartment2 = substr($rucksack, length($rucksack) / 2);
    # println("compartment1: $compartment1");
    # println("compartment2: $compartment2");

    my $elements = [];  # Used to collect items to put in sets.

    # Build sets representing compartments' contents.
    my $i;
    for ($i = 0; $i < length($compartment1); $i++) {
        push(@{$elements}, substr($compartment1, $i, 1));
    }
    my $compartment1_items = make_set(ELEMENTS => $elements);

    # print_hash(
    #     HASH_TO_PRINT => $compartment1_items,
    #     HASH_NAME => "compartment1_items"
    # );

    $elements = [];
    for ($i = 0; $i < length($compartment2); $i++) {
        push(@{$elements}, substr($compartment2, $i, 1));
    }
    my $compartment2_items = make_set(ELEMENTS => $elements);

    # print_hash(
    #     HASH_TO_PRINT => \%compartment2_items,
    #     HASH_NAME => "compartment2_items"
    # );

    # Get all items in common between both compartments.
    my $common_items = intersection(
        SET_1 => $compartment1_items,
        SET_2 => $compartment2_items
    );

    # Calculate total priority of common items.
    for my $k (keys %{$common_items}) {
        $total_priorities += $priorities{$k};
    }
}

println("PART 1");
println("======");
println("Total priorities: $total_priorities");
println("======");

$total_priorities = 0;
my $i;
for ($i = 0; $i <= scalar(@rucksacks) - 3; $i += 3) {
    println("i: $i");
    my $rucksack1_items = make_rucksack_set(RUCKSACK_STR => $rucksacks[$i]);
    my $rucksack2_items = make_rucksack_set(RUCKSACK_STR => $rucksacks[$i + 1]);
    my $rucksack3_items = make_rucksack_set(RUCKSACK_STR => $rucksacks[$i + 2]);

    my $common_items = intersection(
        SET_1 => $rucksack1_items,
        SET_2 => intersection(
            SET_1 => $rucksack2_items,
            SET_2 => $rucksack3_items
        )
    );

    # Calculate total priority of common items.
    for my $k (keys %{$common_items}) {
        $total_priorities += $priorities{$k};
    }
}

println("PART 2");
println("======");
println("Total priorities: $total_priorities");
